internal/util/get/get.go

```diff
@@ -67,41 +68,142 @@ func (c Command) Run() error {
 	// define where we are going to clone the package from
 	r := &git.RepoSpec{OrgRepo: c.Repo, Path: c.Directory, Ref: c.Ref}
 
-	defaultRef, err := gitutil.DefaultRef(c.Repo)
+	err := cloneAndCopy(r, c.Destination, c.Name, c.Clean)
 	if err != nil {
 		return err
 	}
 
-	// clone the repo to a tmp directory.
-	// delete the tmp directory later.
-	err = ClonerUsingGitExec(r, defaultRef)
+	if err = c.fetchRemoteSubpackages(); err != nil {
+		return errors.Wrap(err)
+	}
+	return nil
+}
+
+// fetchRemoteSubpackages goes through the root package and its subpackages
+// and fetches any remote subpackages referenced. It will also handle situations
+// where a remote subpackage references other remote subpackages.
+func (c Command) fetchRemoteSubpackages() error {
+	// Create a stack to keep track of all Kptfiles that needs to be checked
+	// for remote subpackages.
+	stack := newStack()
+
+	paths, err := pathutil.DirsWithFile(c.Destination, kptfile.KptFileName, true)
+	if err != nil {
+		return err
+	}
+	for _, p := range paths {
+		stack.push(p)
+	}
+
+	for stack.len() > 0 {
+		p := stack.pop()
+		kf, err := kptfileutil.ReadFile(p)
+		if err != nil {
+			return err
+		}
+
+		remoteSubPkgDirs := make(map[string]bool)
+		for i := range kf.Subpackages {
+			sp := kf.Subpackages[i]
+
+			if _, found := remoteSubPkgDirs[sp.LocalDir]; found {
+				return fmt.Errorf("multiple remote subpackages with localDir %q", sp.LocalDir)
+			}
+			remoteSubPkgDirs[sp.LocalDir] = true
+
+			gitInfo := sp.Git
+			localPath := filepath.Join(p, sp.LocalDir)
+
+			_, err = os.Stat(localPath)
+			if err != nil && !os.IsNotExist(err) {
+				return err
+			}
+			if !os.IsNotExist(err) {
+				return fmt.Errorf("local subpackage in directory %q already exist", sp.LocalDir)
```

Comment:
Do we ever reach here ? 118-120 lines already cover this. And can this error message be more clear, suggesting the fix as well.

Reply:
Lines 118-120 covers the situation where we get an error that isn't signaling that the folder doesn't exist. We want to handle this separately as this covers unexpected errors. This handles the situation where the subdirectory already exists. This is an "expected" error condition, so we want to provide a better error message.

Added some more context to the error message. I want to revisit how we handle errors separately to see if we should rely on custom types for errors so we can provide better error messages to users.

Reply:
Is the check in line 121 same as `if err == nil && !os.IsNotExist(err)` because the previous check if `if err != nil && !os.IsNotExist(err)`. So if `err == nil` then why would we enter into line 122 ?

Reply:
The second case is already covered under case 1. Consider an error which signals folder exists. It is already hit in line 118.

Reply:
```
if !os.IsNotExist(err) {
				return fmt.Errorf("local subpackage in directory %q already exists. Either"+
					"rename the local subpackage or use a different directory for the remote subpackage", sp.LocalDir)
			}
if err != nil {
  return err
}
```
does what you said.

Reply:
So the code snippet is not the same as the code in the PR and in fact does what I want to avoid. It will report the custom error message for all errors except `IsNotExist`. It will also return an error if the folder doesn't exist, which is the opposite of what we want. We want to return an error if the folder does exist.

I have updated the PR with some comments and changed the check to hopefully make it clearer.